dev/src/reference.ts

```diff
@@ -2699,6 +2743,115 @@ export class Query<
+            });
+            stream.on('data', response => {
+              if (response.document) {
+                readTime = Timestamp.fromProto(response.readTime);
+                const document = this.firestore.snapshot_(
+                  response.document,
+                  response.readTime
+                );
+                const finalDoc = new DocumentSnapshotBuilder<
+                  AppModelType,
+                  DbModelType
+                >(document.ref.withConverter(this._queryOptions.converter));
+                // Recreate the QueryDocumentSnapshot with the DocumentReference
+                // containing the original converter.
+                finalDoc.fieldsProto = document._fieldsProto;
+                finalDoc.readTime = document.readTime;
+                finalDoc.createTime = document.createTime;
+                finalDoc.updateTime = document.updateTime;
+                docs.push(
+                  finalDoc.build() as QueryDocumentSnapshot<
+                    AppModelType,
+                    DbModelType
+                  >
+                );
+              }
+              if (response.stats) {
+                if (response.stats.queryPlan?.planInfo) {
+                  planInfo =
+                    this.firestore._serializer!.decodeGoogleProtobufStruct(
+                      response.stats.queryPlan.planInfo
+                    );
+                }
+                if (response.stats.queryStats) {
+                  executionStats =
+                    this.firestore._serializer!.decodeGoogleProtobufStruct(
+                      response.stats.queryStats
+                    );
+                }
+              }
+              if (response.done) {
+                stream.end();
+              }
+            });
+            stream.on('end', () => {
+              if (this._queryOptions.limitType === LimitType.Last) {
+                // The results for limitToLast queries need to be flipped since
+                // we reversed the ordering constraints before sending the query
+                // to the backend.
+                docs.reverse();
+              }
+              const querySnapshot = new QuerySnapshot(
+                this,
+                readTime,
+                docs.length,
+                () => docs,
+                () => {
+                  const changes: Array<
+                    DocumentChange<AppModelType, DbModelType>
+                  > = [];
+                  for (let i = 0; i < docs.length; ++i) {
+                    changes.push(new DocumentChange('added', docs[i], -1, i));
+                  }
+                  return changes;
+                }
+              );
+              resolve(
+                new QueryProfileInfo<QuerySnapshot<AppModelType, DbModelType>>(
+                  planInfo,
+                  executionStats,
+                  querySnapshot
+                )
+              );
+            });
```

Comment:
I agree that this rewrite of `_getQueryProfileInfo(...)` is simpler than using `_stream(...)` and all of that baggage. Nice change here.

nit: It's unfortunate that there are two implementations of creating a QuerySnapshot between `get()` and `explainAnalyze()`. That brings some risk of deviation in these two implementations in the future.

Reply:
Agreed. IMO to address the risk of deviation, we first need to do some cleanups in `_stream(...)`, and then refactor the two functions into one. I'll file a ticket for this work.